### PR TITLE
Refactor source mode and add custom javascript worker

### PR DESCRIPTION
**Pull request by BobbyZhouZijian**
* refactored some code

* add a custom javascript worker as we are migrating to use ace-builder for the editor.

### CODE REVIEW

File: src/editors/ace/modes/source.ts [hunk 1]
```diff
@@ -122,7 +121,7 @@ export function HighlightRulesSelector(id: number) {
 
           'storage.type': 'const|let|function',
 
-          'support.function': 'alert|' + ChapterFunctionNameSelector(),
+          'support.function': ChapterFunctionNameSelector(),
```
Comment (martin-henz) [inline src/editors/ace/modes/source.ts]: nice catch

File: src/editors/ace/worker/javascript.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
...dash.isArguments=isArguments,lodash.isArray=isArray,lodash.isEmpty=isEmpty,lodash.isFunction=isFunction,lodash.isNative=isNative,lodash.isNumber=isNumber,lodash.isObject=isObject,lodash.isPlainObject=isPlainObject,lodash.isString=isString,lodash.isTypedArray=isTypedArray,lodash.last=last,lodash.some=some,lodash.any=some,lodash.contains=includes,lodash.include=includes,lodash.VERSION=VERSION,freeExports&&freeModule?moduleExports?(freeModule.exports=lodash)._=lodash:freeExports._=lodash:root._=lodash\n}).call(this)}).call(this,"undefined"!=typeof global?global:"undefined"!=typeof self?self:"undefined"!=typeof window?window:{})},{}],"/node_modules/jshint/src/jshint.js":[function(_dereq_,module,exports){var _=_dereq_("../lodash"),events=_dereq_("events"),vars=_dereq_("./vars.js"),messages=_dereq_("./messages.js"),Lexer=_dereq_("./lex.js").Lexer,reg=_dereq_("./reg.js"),state=_dereq_("./state.js").state,style=_dereq_("./style.js"),options=_dereq_("./options.js"),scopeManager=_dereq_("./scope-manager.js"),JSHINT=function(){"use strict";function checkOption(name,t){return name=name.trim(),/^[+-]W\\d{3}$/g.test(name)?!0:-1!==options.validNames.indexOf(name)||"jslint"===t.type||_.has(options
```
Comment (martin-henz) [inline src/editors/ace/worker/javascript.ts]: Where does this code come from? Why is it in one line? How would we be able to maintain this?

Comment (BobbyZhouZijian) [reply src/editors/ace/worker/javascript.ts]: I took it directly from the brace library. I will try if i can format the script.

Comment (BobbyZhouZijian) [reply src/editors/ace/worker/javascript.ts]: can't seem to think of better solution now

Comment (martin-henz) [reply src/editors/ace/worker/javascript.ts]: Can you add a comment that describes where this code is from, and how to generate it? What if there is a bug in that code?

Comment (BobbyZhouZijian) [reply src/editors/ace/worker/javascript.ts]: ok sure.

Comment (BobbyZhouZijian) [reply src/editors/ace/worker/javascript.ts]: have added comments

**Comment by BobbyZhouZijian** [comment]
we may need to bump the js-slang version

**Pull request merged**